File: ai_assist_context_service/redaction.py
import re

from .validation import assert_non_empty_string

REDACTION_POLICIES = {
    "NONE": "none",
    "MVP_DEFAULT": "mvp_default",
}

DEFAULT_REDACTION_RULES = (
    {
        "id": "email_address",
        "pattern": re.compile(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", re.IGNORECASE),
        "replacement": "<redacted:email_address>",
    },
    {
        "id": "api_key_like_secret",
        "pattern": re.compile(r"\b(?:sk|pk|api|token|secret)[_-][A-Za-z0-9_-]{16,}\b"),
        "replacement": "<redacted:secret>",
    },
    {
        "id": "bearer_token",
        "pattern": re.compile(r"\bBearer\s+[A-Za-z0-9._~+/=-]{20,}\b", re.IGNORECASE),
        "replacement": "Bearer <redacted:token>",
    },
)


def apply_context_redaction(content, options=None):
    options = options or {}
    assert_non_empty_string(content, "content")
    policy = options.get("policy", REDACTION_POLICIES["NONE"])
    if policy is None:
        policy = REDACTION_POLICIES["NONE"]

    if policy == REDACTION_POLICIES["NONE"]:
        return {
            "content": content,
            "metadata": {
                "policy": policy,
                "redacted": False,
                "rulesApplied": [],
            },
        }

    if policy != REDACTION_POLICIES["MVP_DEFAULT"]:
        raise TypeError(f"Unsupported redaction policy: {policy}")

    rules = options.get("rules") or DEFAULT_REDACTION_RULES
    redacted_content = content
    rules_applied = []

    for rule in rules:
        next_content = rule["pattern"].sub(rule["replacement"], redacted_content)
        if next_content != redacted_content:
            rules_applied.append(rule["id"])
            redacted_content = next_content

    return {
        "content": redacted_content,
        "metadata": {
            "policy": policy,
            "redacted": len(rules_applied) > 0,
            "rulesApplied": rules_applied,
        },
    }

File: src/redaction.js
import { assertNonEmptyString } from "./validation.js";

export const REDACTION_POLICIES = Object.freeze({
  NONE: "none",
  MVP_DEFAULT: "mvp_default"
});

const DEFAULT_REDACTION_RULES = Object.freeze([
  {
    id: "email_address",
    pattern: /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/giu,
    replacement: "<redacted:email_address>"
  },
  {
    id: "api_key_like_secret",
    pattern: /\b(?:sk|pk|api|token|secret)[_-][A-Za-z0-9_-]{16,}\b/gu,
    replacement: "<redacted:secret>"
  },
  {
    id: "bearer_token",
    pattern: /\bBearer\s+[A-Za-z0-9._~+/=-]{20,}\b/giu,
    replacement: "Bearer <redacted:token>"
  }
]);

export function applyContextRedaction(content, options = {}) {
  assertNonEmptyString(content, "content");
  const policy = options.policy ?? REDACTION_POLICIES.NONE;

  if (policy === REDACTION_POLICIES.NONE) {
    return {
      content,
      metadata: {
        policy,
        redacted: false,
        rulesApplied: []
      }
    };
  }

  if (policy !== REDACTION_POLICIES.MVP_DEFAULT) {
    throw new TypeError(`Unsupported redaction policy: ${policy}`);
  }

  const rules = options.rules ?? DEFAULT_REDACTION_RULES;
  let redactedContent = content;
  const rulesApplied = [];

  for (const rule of rules) {
    const nextContent = redactedContent.replace(rule.pattern, rule.replacement);
    if (nextContent !== redactedContent) {
      rulesApplied.push(rule.id);
      redactedContent = nextContent;
    }
  }

  return {
    content: redactedContent,
    metadata: {
      policy,
      redacted: rulesApplied.length > 0,
      rulesApplied
    }
  };
}
